src/jit/codegenxarch.cpp

```diff
@@ -5001,6 +5001,20 @@ void CodeGen::genCallInstruction(GenTreePtr node)
 
 #endif // defined(_TARGET_X86_)
 
+#ifdef FEATURE_AVX_SUPPORT
+    // When it's a PInvoke call and the call type is USER function, we issue VZEROUPPER here
+    // if the function contains 256bit AVX instructions, this is to avoid AVX-256 to Legacy SSE
+    // transition penalty, assuming the user function contains legacy SSE instruction
+    if (call->IsPInvoke() && call->gtCallType == CT_USER_FUNC &&
+        compiler->getFloatingPointInstructionSet() == InstructionSet_AVX)
```

Comment:
We don't need to check instruction set here because emitter flag is set based on instruction set.  Instead it would suffice to assert here as follows

if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) && getEmitter()->Contains256bitAVX())
{
     assert(compiler->getFloatingPointInstructionSet() == InstructionSet_AVX);
     instGen(INS_vzeroupper);
}

---
In reply to: [95471453](https://github.com/dotnet/coreclr/pull/8588#discussion_r95471453) [](ancestors = 95471453)